lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -30,36 +30,112 @@ using namespace firrtl;
 // Node class
 //===----------------------------------------------------------------------===//
 
-using CombPathsType = SmallVector<SmallVector<size_t, 2>>;
-using CombPathsMap = DenseMap<Operation *, CombPathsType>;
-
 using ConnectIterator =
     mlir::detail::op_iterator<FConnectLike, Region::OpIterator>;
 using ConnectRange = llvm::iterator_range<ConnectIterator>;
+// Pair to represent port number and field index.
+using PortField = std::pair<long, long>;
+// Map to represent the combinational paths from a PortField to any other
+// PortFields.
+using CombPathsType = DenseMap<PortField, SmallVector<PortField>>;
+// Map to represent combinational paths between all the ports of a module.
+using CombPathsMap = DenseMap<Operation *, CombPathsType>;
 
 namespace {
 /// The graph context containing pointers of the combinational paths map and the
 /// instance graph.
 struct NodeContext {
-  CombPathsMap *map;
+  CombPathsMap &map;
   InstanceGraph *graph;
   ConnectRange connects;
-
-  explicit NodeContext(CombPathsMap *map, InstanceGraph *graph,
-                       ConnectRange connects)
-      : map(map), graph(graph), connects(connects) {}
+  // Record the last visited port that is source to a connect.
+  PortField srcPort;
+  // Current module being procesed.
+  FModuleOp module;
+  // All the comb paths between the current module's ports.
+  CombPathsType *portPaths;
+  // Set true, when the isCycle() is executing. Required to ensure certain data
+  // structures are not updated during the Graph traversal inside the isCycle
+  // call.
+  bool cycleCheck;
+
+  explicit NodeContext(CombPathsMap &map, InstanceGraph *graph,
+                       ConnectRange connects, FModuleOp module)
+      : map(map), graph(graph), connects(connects), module(module) {
+    // Cache the ports path for this module. This ensures, that the current
+    // module's paths are updated in `map`.
+    portPaths = &map[module];
```

Comment:
I'm wondering this `portPaths` can be freed when `map` reallocates elements. Do you happen to know this could happen?